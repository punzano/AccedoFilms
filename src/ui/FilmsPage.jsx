import React, { useEffect, useRef } from 'react';
import { Link, useNavigate } from 'react-router-dom';

import useHorizontalWheelScroll from './hooks/useHorizontalWheelScroll';

const FilmsPage = (props) => {
    const { dispatch, films, getFilms } = props;
    const { list: filmsList } = films;

    const navigate = useNavigate();

    useEffect(() => {
        dispatch(getFilms());
    }, [dispatch, getFilms]);

    // This doesn't work, it's objective would have been to perform the scroll with the mouse wheel,
    // but I don't have enough time to check what is wrong
    const [ref] = useHorizontalWheelScroll();

    const onFilmClicked = (ev) => {
        const videoId = ev.currentTarget.id;
        navigate(`/video/${videoId}`);
    };

    const getFilmsElements = (filmsList) => (
        filmsList.map(film => (
            <div ref={ref} key={film.id} id={film.id} className="film-container" onClick={onFilmClicked}>
                <img className="film-image" src={film.image} alt={film.name}/>
                <span className="film-title" data-testid="film-title">{film.name}</span>
                <div className="description-overlay">
                    <p className="description-text">
                        {film.summary}
                    </p>
                </div>
            </div>
        ))
    );

    return (
        <div className="films-page-container" data-testid="films-page-container">
            <header className="header">
                <h1 className="app-title">Welcome to Accedo Films</h1>
                <Link to="/user-history">User history</Link>
            </header>
            <section className="films-container">
                {filmsList.length > 0 && getFilmsElements(filmsList)}
            </section>
        </div>
    );
};

export default FilmsPage;