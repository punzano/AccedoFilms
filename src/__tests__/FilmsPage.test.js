import '@babel/polyfill';
import React from 'react';
import { render, fireEvent, cleanup, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history'
import FilmsPage from '../ui/FilmsPage';
import './mocks/localStorage';

afterEach(cleanup);

const mockedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigate
}));

const mockProps = {
    dispatch: jest.fn(),
    films: {
        list: [
            {
                "id": 2259634,
                "name": "The Bump 139",
                "season": 4,
                "number": 5,
                "type": "regular",
                "image": "https://static.tvmaze.com/uploads/images/medium_portrait/217/542886.jpg",
                "mediaUrl": "https://test-data-interviews.s3.eu-west-1.amazonaws.com/bbb_multiaudio.mp4",
                "summary": "WWE's The Bump is unlike any WWE show you've ever seen before, featuring a dynamic cast of hosts--led by Kayla Braxton--and WWE Superstars stopping in weekly, both as in-studio and video call-in guests."
              },
              {
                "id": 2262125,
                "name": "Lisa Ling, Amber Ruffin",
                "season": 1,
                "number": 2,
                "image": "https://static.tvmaze.com/uploads/images/medium_portrait/359/898286.jpg",
                "mediaUrl": "https://test-data-interviews.s3.eu-west-1.amazonaws.com/Forest+-+97998.mp4",
                "summary": "Lisa Ling guest co-hosts; TV host Amber Ruffin."
              }
        ]
    }
};

const mockFilmsActions = {
    getFilms: jest.fn(),
};

global.window = { location: { pathname: null } };

describe('FilmPage component', () => {
    test('FilmPage snapshot', async () => {
        render(
            <BrowserRouter>
                <FilmsPage {...mockProps} {...mockFilmsActions}/>
            </BrowserRouter>
        );
        const filmsPageContainer = screen.getByTestId('films-page-container');
        
        expect(filmsPageContainer).toMatchSnapshot();
    });

    test('FilmPage should call dispatch', async () => {
        render(
            <BrowserRouter>
                <FilmsPage {...mockProps} {...mockFilmsActions}/>
            </BrowserRouter>
        );
        expect(mockProps.dispatch).toHaveBeenCalled();
        expect(mockFilmsActions.getFilms).toHaveBeenCalled();
    });

    test('FilmPage should render 2 films', async () => {
        render(
            <BrowserRouter>
                <FilmsPage {...mockProps} {...mockFilmsActions}/>
            </BrowserRouter>
        );
        
        const filmsName = screen.getAllByTestId('film-title');
        expect(filmsName.length).toBe(2);
        expect(filmsName[0].textContent).toBe('The Bump 139');
        expect(filmsName[1].textContent).toBe('Lisa Ling, Amber Ruffin');
    });

    test('FilmPage should navigate to another url if a film is clicked', async () => {
        const history = createMemoryHistory();
        render(
            <BrowserRouter history={history}>
                <FilmsPage {...mockProps} {...mockFilmsActions}/>
            </BrowserRouter>
        );
        
        const filmsPageContainer = screen.getByTestId('films-page-container');
        fireEvent.click(filmsPageContainer);
        setTimeout(() => {
            expect(mockedNavigate).toHaveBeenCalledWith('/video/2259634');
        }, 1000);
        //The optimal approach would be somthing like this, but it doesn't work as expected
        /* await waitFor(() => {
            expect(mockedNavigate).toHaveBeenCalledWith('/video/2259634');
        }); */     
    });
});
