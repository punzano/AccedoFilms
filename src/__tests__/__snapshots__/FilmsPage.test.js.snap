// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmPage component FilmPage snapshot 1`] = `
<div
  class="films-page-container"
  data-testid="films-page-container"
>
  <header
    class="header"
  >
    <h1
      class="app-title"
    >
      Welcome to Accedo Films
    </h1>
    <a
      href="/user-history"
    >
      User history
    </a>
  </header>
  <section
    class="films-container"
  >
    <div
      class="film-container"
      id="2259634"
    >
      <img
        alt="The Bump 139"
        class="film-image"
        src="https://static.tvmaze.com/uploads/images/medium_portrait/217/542886.jpg"
      />
      <span
        class="film-title"
        data-testid="film-title"
      >
        The Bump 139
      </span>
      <div
        class="description-overlay"
      >
        <p
          class="description-text"
        >
          WWE's The Bump is unlike any WWE show you've ever seen before, featuring a dynamic cast of hosts--led by Kayla Braxton--and WWE Superstars stopping in weekly, both as in-studio and video call-in guests.
        </p>
      </div>
    </div>
    <div
      class="film-container"
      id="2262125"
    >
      <img
        alt="Lisa Ling, Amber Ruffin"
        class="film-image"
        src="https://static.tvmaze.com/uploads/images/medium_portrait/359/898286.jpg"
      />
      <span
        class="film-title"
        data-testid="film-title"
      >
        Lisa Ling, Amber Ruffin
      </span>
      <div
        class="description-overlay"
      >
        <p
          class="description-text"
        >
          Lisa Ling guest co-hosts; TV host Amber Ruffin.
        </p>
      </div>
    </div>
  </section>
</div>
`;
